### PR TITLE
Fix NotifyQt::askForPassword QInputDialog is not created in GUI Thread.

### PR DESCRIPTION
**Pull request by PhenomRetroShare**
To test without it, change in:
/retroshare-gui/src/gui/GenCertDialog.cpp:665
QTimer::singleShot(30000, []() { rsNotify->clearPgpPassphrase(); } );
to
QTimer::singleShot(**100**, []() { rsNotify->clearPgpPassphrase(); } );

And create a new node.

### CODE REVIEW

File: retroshare-gui/src/gui/MainWindow.cpp [hunk 1]
```diff
@@ -1688,3 +1689,21 @@ void MainWindow::setCompactStatusMode(bool compact)
 	ratesstatus->setCompactMode(compact);
 	//opModeStatus: TODO Show only ???
 }
+
+// Called by other thread because Widgets must be created in the GUI thread.
+Gui_InputDialogReturn MainWindow::guiInputDialog(const QString& windowTitle, const QString& labelText, QLineEdit::EchoMode textEchoMode, bool modal)
+{
+
+	RsErr(this->thread(), " <-> ", qApp->thread());
```
Comment (G10h4ck) [inline retroshare-gui/src/gui/MainWindow.cpp]: AFAIU you should use `RsDbg` or `RsInfo` here not `RsErr`

File: libretroshare/src/pgp/pgphandler.cc [hunk 2]
```diff
@@ -1381,10 +1381,12 @@ bool PGPHandler::decryptTextFromFile(const RsPgpId&,std::string& text,const std:
 
 bool PGPHandler::SignDataBin(const RsPgpId& id,const void *data, const uint32_t len, unsigned char *sign, unsigned int *signlen,bool use_raw_signature, std::string reason /* = "" */)
 {
-	RsStackMutex mtx(pgphandlerMtx) ;				// lock access to PGP memory structures.
 	// need to find the key and to decrypt it.
-	
-	const ops_keydata_t *key = locked_getSecretKey(id) ;
+	const ops_keydata_t *key = nullptr;
+	{
+		RsStackMutex mtx(pgphandlerMtx) ;				// lock access to PGP memory structures.
```
Comment (G10h4ck) [inline libretroshare/src/pgp/pgphandler.cc]: better use the `RS_STACK_MUTEX` macro so we avoid compile complaining and better debug is possible

File: libretroshare/src/pgp/pgphandler.cc [hunk 3]
```diff
@@ -1404,35 +1406,38 @@ bool PGPHandler::SignDataBin(const RsPgpId& id,const void *data, const uint32_t
...
+	}
+	if(!secret_key)
+	{
+		std::cerr << "Could not obtain secret key. Signature cancelled." << std::endl;
+		return false ;
+	}
 
 	// then do the signature.
 
+	RsStackMutex mtx(pgphandlerMtx) ;				// lock access to PGP memory structures.
```
Comment (G10h4ck) [inline libretroshare/src/pgp/pgphandler.cc]: better use the RS_STACK_MUTEX macro so we avoid compile complaining and better debug is possible

File: retroshare-gui/src/gui/MainWindow.cpp (hunk 1 above)
Comment (PhenomRetroShare) [reply retroshare-gui/src/gui/MainWindow.cpp]: Done

File: libretroshare/src/pgp/pgphandler.cc (hunk 2 above)
Comment (PhenomRetroShare) [reply libretroshare/src/pgp/pgphandler.cc]: Done

File: libretroshare/src/pgp/pgphandler.cc (hunk 3 above)
Comment (PhenomRetroShare) [reply libretroshare/src/pgp/pgphandler.cc]: Done

**Comment by PhenomRetroShare** [comment]
@csoler ;)

File: libretroshare/src/pgp/pgphandler.cc [hunk 4]
```diff
@@ -1381,10 +1381,12 @@ bool PGPHandler::decryptTextFromFile(const RsPgpId&,std::string& text,const std:
 
 bool PGPHandler::SignDataBin(const RsPgpId& id,const void *data, const uint32_t len, unsigned char *sign, unsigned int *signlen,bool use_raw_signature, std::string reason /* = "" */)
 {
-	RsStackMutex mtx(pgphandlerMtx) ;				// lock access to PGP memory structures.
```
Comment (csoler) [inline libretroshare/src/pgp/pgphandler.cc]: Moving this mutex is really not a good idea: locked_getSecretKey(id) returns a pointer in the middle of a large array containing all key structures. If anything happens to the keyring between the getSecretKey and the use of it, then an undefined behavior will happen.

File: libretroshare/src/pgp/pgphandler.cc [hunk 5]
```diff
@@ -1381,10 +1381,12 @@ bool PGPHandler::decryptTextFromFile(const RsPgpId&,std::string& text,const std:
...
 	// need to find the key and to decrypt it.
-	
-	const ops_keydata_t *key = locked_getSecretKey(id) ;
+	const ops_keydata_t *key = nullptr;
+	{
+		RS_STACK_MUTEX(pgphandlerMtx); // lock access to PGP memory structures.
+		key = locked_getSecretKey(id);
```
Comment (csoler) [inline libretroshare/src/pgp/pgphandler.cc]: If you really need that mutex here, then copy the struct entirely. It's costly, but that shouldn't be a problem here.

Comment (PhenomRetroShare) [reply libretroshare/src/pgp/pgphandler.cc]: Done

**Pull request merged**